File: macos/Sources/ClaudeStatusCore/StatusStore.swift
import Foundation

/// State of a single Claude session (traffic-light color).
public enum State: String {
    case red    // asking a question / waiting for permission — waiting on you
    case yellow // working
    case green  // finished, your turn

    /// Priority: higher = needs more attention. Used when aggregating.
    public var priority: Int {
        switch self {
        case .red: return 3
        case .yellow: return 2
        case .green: return 1
        }
    }

    public var emoji: String {
        switch self {
        case .red: return "🔴"
        case .yellow: return "🟡"
        case .green: return "🟢"
        }
    }
}

/// Status record of a single Claude session (maps to one status file).
public struct SessionStatus {
    public let sessionID: String
    public let state: State
    public let project: String
    public let cwd: String
    public let ts: Date
    public let platform: String // "desktop" | "vscode" | "cursor" | "terminal" | "unknown"
    public let appPath: String  // path of the hosting .app (if any) — used to focus it on click
    public let pid: Int32       // the running `claude` process (0 if unknown) — used to end the session

    public init(sessionID: String, state: State, project: String,
                cwd: String, ts: Date, platform: String, appPath: String = "", pid: Int32 = 0) {
        self.sessionID = sessionID
        self.state = state
        self.project = project
        self.cwd = cwd
        self.ts = ts
        self.platform = platform
        self.appPath = appPath
        self.pid = pid
    }
}

/// Reads and aggregates the `~/.claude/status/*.json` files.
public final class StatusStore {

    /// Records older than this are considered stale (stuck session) and hidden.
    public static let staleAfter: TimeInterval = 30 * 60 // 30 min

    public let statusDir: URL

    /// When `statusDir` is nil, `~/.claude/status` is used. Parameterized so tests
    /// can inject a temporary directory.
    public init(statusDir: URL? = nil) {
        if let dir = statusDir {
            self.statusDir = dir
        } else {
            self.statusDir = FileManager.default.homeDirectoryForCurrentUser
                .appendingPathComponent(".claude", isDirectory: true)
                .appendingPathComponent("status", isDirectory: true)
        }
    }

    /// Reads all valid (non-stale) sessions in the directory.
    public func load() -> [SessionStatus] {
        let fm = FileManager.default
        guard let entries = try? fm.contentsOfDirectory(
            at: statusDir,
            includingPropertiesForKeys: nil,
            options: [.skipsHiddenFiles]
        ) else {
            return []
        }

        let now = Date()
        var results: [SessionStatus] = []

        for url in entries where url.pathExtension == "json" {
            guard
                let data = try? Data(contentsOf: url),
                let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                let stateRaw = obj["state"] as? String,
                let state = State(rawValue: stateRaw)
            else { continue }

            let tsNumber = (obj["ts"] as? NSNumber)?.doubleValue
                ?? Double(obj["ts"] as? String ?? "")
                ?? 0
            let ts = Date(timeIntervalSince1970: tsNumber)

            // Liveness: if the session's claude process is gone (chat closed), drop
            // it and delete the now-useless file (keep disk clean).
            let pid = (obj["session_pid"] as? NSNumber)?.int32Value ?? 0
            if pid > 0 {
                if !StatusStore.isProcessAlive(pid) {
                    try? fm.removeItem(at: url)
                    continue
                }
            } else {
                // Old format (no pid) → fall back to time-based staleness check.
                if now.timeIntervalSince(ts) > StatusStore.staleAfter {
                    try? fm.removeItem(at: url)
                    continue
                }
            }

            let cwd = obj["cwd"] as? String ?? ""
            let project = (obj["project"] as? String)
                ?? (cwd.isEmpty ? "?" : (cwd as NSString).lastPathComponent)
            let sessionID = url.deletingPathExtension().lastPathComponent

            let platform = (obj["platform"] as? String) ?? "unknown"
            let appPath = (obj["app_path"] as? String) ?? ""

            results.append(SessionStatus(
                sessionID: sessionID,
                state: state,
                project: project,
                cwd: cwd,
                ts: ts,
                platform: platform,
                appPath: appPath,
                pid: pid
            ))
        }

        // Sort by priority (red on top); ties broken by newest first.
        results.sort { a, b in
            if a.state.priority != b.state.priority {
                return a.state.priority > b.state.priority
            }
            return a.ts > b.ts
        }
        return results
    }

    /// Aggregate state shown on the bar icon. Green (idle) when there are no sessions.
    public func aggregate(_ sessions: [SessionStatus]) -> State {
        sessions.map { $0.state }.max { $0.priority < $1.priority } ?? .green
    }

    /// Checks whether the given PID is alive (POSIX `kill(pid, 0)`).
    public static func isProcessAlive(_ pid: Int32) -> Bool {
        if pid <= 0 { return true }
        if kill(pid, 0) == 0 { return true }
        return errno == EPERM // process exists but we lack permission to signal it → alive
    }
}

File: windows/ClaudeTrafficLight/Core/StatusStore.cs
using System.Text.Json;

namespace ClaudeTrafficLight.Core;

/// <summary>Traffic-light state of a single Claude session.</summary>
public enum State
{
    Red,    // asking a question / waiting for permission — waiting on you
    Yellow, // working
    Green   // finished, your turn
}

public static class StateExtensions
{
    /// <summary>Priority: higher = needs more attention. Used when aggregating.</summary>
    public static int Priority(this State s) => s switch
    {
        State.Red => 3,
        State.Yellow => 2,
        State.Green => 1,
        _ => 0
    };

    public static string Emoji(this State s) => s switch
    {
        State.Red => "🔴",
        State.Yellow => "🟡",
        State.Green => "🟢",
        _ => ""
    };

    /// <summary>Parse the JSON "state" string; null if invalid.</summary>
    public static State? Parse(string? raw) => raw switch
    {
        "red" => State.Red,
        "yellow" => State.Yellow,
        "green" => State.Green,
        _ => null
    };
}

/// <summary>Status record of a single Claude session (maps to one status file).</summary>
public sealed record SessionStatus(
    string SessionId,
    State State,
    string Project,
    string Cwd,
    DateTimeOffset Ts,
    string Platform,   // "desktop" | "vscode" | "cursor" | "terminal" | "unknown"
    string AppPath,    // path of the hosting app (if any) — used to focus it on click
    int Pid);          // the running `claude` process (0 if unknown) — used to end the session

/// <summary>Reads and aggregates the <c>%USERPROFILE%\.claude\status\*.json</c> files.</summary>
public sealed class StatusStore
{
    /// <summary>Records older than this are considered stale (stuck session) and hidden.</summary>
    public static readonly TimeSpan StaleAfter = TimeSpan.FromMinutes(30);

    public string StatusDir { get; }

    /// <summary>When <paramref name="statusDir"/> is null, <c>%USERPROFILE%\.claude\status</c> is used.</summary>
    public StatusStore(string? statusDir = null)
    {
        StatusDir = statusDir ?? Paths.StatusDir;
    }

    /// <summary>Reads all valid (non-stale, live) sessions in the directory.</summary>
    public List<SessionStatus> Load()
    {
        var results = new List<SessionStatus>();
        string[] files;
        try
        {
            files = Directory.GetFiles(StatusDir, "*.json");
        }
        catch
        {
            return results;
        }

        var now = DateTimeOffset.UtcNow;

        foreach (var path in files)
        {
            JsonElement obj;
            try
            {
                obj = ReadJsonWithRetry(path);
            }
            catch
            {
                continue; // unreadable / invalid JSON — skip (row reappears next poll)
            }

            if (obj.ValueKind != JsonValueKind.Object) continue;

            var state = StateExtensions.Parse(GetString(obj, "state"));
            if (state is null) continue;

            var ts = ParseTs(obj);

            // Liveness: if the session's claude process is gone (chat closed), drop
            // it and delete the now-useless file (keep disk clean).
            int pid = GetInt(obj, "session_pid");
            if (pid > 0)
            {
                if (!Platform.ProcessInfo.IsAlive(pid))
                {
                    TryDelete(path);
                    continue;
                }
            }
            else
            {
                // Old format (no pid) → fall back to time-based staleness check.
                if (now - ts > StaleAfter)
                {
                    TryDelete(path);
                    continue;
                }
            }

            string cwd = GetString(obj, "cwd") ?? "";
            string project = GetString(obj, "project")
                ?? (string.IsNullOrEmpty(cwd) ? "?" : LastPathComponent(cwd));
            string sessionId = Path.GetFileNameWithoutExtension(path);
            string platform = GetString(obj, "platform") ?? "unknown";
            string appPath = GetString(obj, "app_path") ?? "";

            results.Add(new SessionStatus(sessionId, state.Value, project, cwd, ts, platform, appPath, pid));
        }

        // Sort by priority (red on top); ties broken by newest first.
        results.Sort((a, b) =>
        {
            if (a.State.Priority() != b.State.Priority())
                return b.State.Priority().CompareTo(a.State.Priority());
            return b.Ts.CompareTo(a.Ts);
        });
        return results;
    }

    /// <summary>Aggregate state shown on the bar icon. Green (idle) when there are no sessions.</summary>
    public State Aggregate(IReadOnlyList<SessionStatus> sessions)
    {
        var best = State.Green;
        bool any = false;
        foreach (var s in sessions)
        {
            if (!any || s.State.Priority() > best.Priority()) { best = s.State; any = true; }
        }
        return any ? best : State.Green;
    }

    // MARK: - helpers

    private static string? GetString(JsonElement obj, string key)
        => obj.TryGetProperty(key, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;

    private static int GetInt(JsonElement obj, string key)
    {
        if (!obj.TryGetProperty(key, out var v)) return 0;
        return v.ValueKind switch
        {
            JsonValueKind.Number => v.TryGetInt32(out var n) ? n : 0,
            JsonValueKind.String => int.TryParse(v.GetString(), out var n) ? n : 0,
            _ => 0
        };
    }

    private static DateTimeOffset ParseTs(JsonElement obj)
    {
        double secs = 0;
        if (obj.TryGetProperty("ts", out var v))
        {
            if (v.ValueKind == JsonValueKind.Number && v.TryGetDouble(out var d)) secs = d;
            else if (v.ValueKind == JsonValueKind.String && double.TryParse(v.GetString(), out var ds)) secs = ds;
        }
        return DateTimeOffset.FromUnixTimeSeconds((long)secs);
    }

    private static string LastPathComponent(string p)
    {
        p = p.TrimEnd('/', '\\');
        int i = p.LastIndexOfAny(new[] { '/', '\\' });
        return i >= 0 ? p[(i + 1)..] : p;
    }

    /// <summary>
    /// Reads and parses a status file, retrying once after a short beat on IO errors:
    /// the hook's ReplaceFile can transiently lock the file, and skipping it would make
    /// the session row blink out of the list for a whole poll cycle.
    /// </summary>
    private static JsonElement ReadJsonWithRetry(string path)
    {
        for (int i = 0; ; i++)
        {
            try
            {
                using var doc = JsonDocument.Parse(File.ReadAllText(path));
                return doc.RootElement.Clone();
            }
            catch (IOException) when (i < 1)
            {
                Thread.Sleep(15);
            }
        }
    }

    /// <summary>Deletes a dead session's status file together with its hook-side
    /// <c>.lock</c> companion (the hook serializes concurrent writers through it;
    /// when the session dies without a SessionEnd, only we ever clean that up).</summary>
    private static void TryDelete(string path)
    {
        try { File.Delete(path); } catch { /* ignore */ }
        try { File.Delete(path + ".lock"); } catch { /* ignore */ }
    }
}
